LFS-842: Improve UX in the Variants Upload form

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-842) 

Various fixes to the UI of the Variants uploader.

![image](https://user-images.githubusercontent.com/651980/116155773-93468e00-a6b8-11eb-8f90-1a0c0c714438.png)

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Hide upload button when no file is selected.

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Add the top instructions

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Remove green colouring and "File", improve layout slightly

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Hide other versions by default

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Switch the "show other versions" to a dialog with MaterialTable

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Fix the title line

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -718,60 +732,106 @@ export default function VariantFilesContainer() {
-                  { file?.region?.path && <span>Region nb: <Link href={regionPath} target="_blank"> {file.region.id} </Link> </span> }
-                  { file.formPath && <span>Form: <Link href={file.formPath.replace("/Forms", "Forms")} target="_blank"> {file.formPath.replace("/Forms/", "")} </Link></span> }
+                <Typography variant="overline" className={classes.fileDetail}>
+                  Patient <Link href={subjectPath} target="_blank"> {file.subject.id} </Link> /
```
Use the subject type label

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -718,60 +732,106 @@ export default function VariantFilesContainer() {
+                  Patient <Link href={subjectPath} target="_blank"> {file.subject.id} </Link> /
+                  Tumor <Link href={tumorPath} target="_blank"> {file.tumor.id} </Link>
+                  { file?.region?.path && <> / Tumor region: <Link href={regionPath} target="_blank"> {file.region.id} </Link> </> }
+                  { file.formPath && <span>Somatic Variants / <Link href={file.formPath.replace("/Forms", "Forms")} target="_blank"> {file.name} </Link></span> }
```
Use the questionnaire title

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -665,6 +672,12 @@ export default function VariantFilesContainer() {
           onSubmit={upload}
           key="file-upload">
       <Typography component="h2" variant="h5" className={classes.dialogTitle}>Variants Upload</Typography>
+      <Typography>Please upload files named according to one of the following formats:</Typography>
+      <ul>
+        <li><i>PatientId_TumorNumber.csv (e.g. AB12345_1.csv)</i></li>
+        <li><i>PatientId_TumorNumber_TumorRegion.csv (e.g. AB12345_1_a.csv)</i></li>
+      </ul>
+      <Typography>The file will be automatically associated with the subject identified by its name. You can review and correct the identifiers after selecting the file, before upload.</Typography>
```
```suggestion
      <p>Please upload files named according to one of the following formats:</p>
      <ul>
        <li><i>PatientId_TumorNumber.csv</i> (e.g. <i>AB12345_1.csv</i>)</li>
        <li><i>PatientId_TumorNumber_TumorRegion.csv</i> (e.g. <i>AB12345_1_a.csv</i>)</li>
      </ul>
      <p>The file will be automatically associated with the subject identified by its name. You can review and correct the identifiers after selecting the file, before upload.</p>
```

marta- replied:
Also, there should be more spacing before and after the `dropzone` (`theme.spacing(3)`).

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -681,21 +694,22 @@ export default function VariantFilesContainer() {
-          </span>
-        </Button>
+        { selectedFiles.length ?
+          <Button type="submit" variant="contained" color="primary" disabled={uploadInProgress || !!error && selectedFiles.length == 0} className={classes.uploadButton}>
```
The button should be in line with the extracted id inputs (see 2nd mockup in the task).

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -718,60 +732,106 @@ export default function VariantFilesContainer() {
-    </span> }
+    </span>
+    }
+  <Dialog open={showVersionsDialog} onClose={() => setShowVersionsDialog(false)}>
```
There's no visible way to get out of the dialog. Please add a (x) close button in the DialogTitle. Also the dialog is a bit too narrow, let's give it a minWidth.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -718,60 +732,106 @@ export default function VariantFilesContainer() {
+                    ?
+                      <p>There are no versions of this file.</p>
+                    :
+                      <Link
```
The link should be inside the `p`, not the other way around. Also:
- the `caption` variant should be used
- the texts looks misaligned because it has a left padding it shouldn't

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -665,6 +672,12 @@ export default function VariantFilesContainer() {
           onSubmit={upload}
           key="file-upload">
       <Typography component="h2" variant="h5" className={classes.dialogTitle}>Variants Upload</Typography>
```
```suggestion
      <Typography variant="h2">Variants Upload</Typography>
```

OrdiNeu replied:
Fixed in `ef18a81` :heavy_check_mark:

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Fix the close button and several other parts of formatting

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Move the button outside to the inputs

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Make a minimum width on the variants upload previous version dialog, and
remove an unused import.

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Increase the spacing around the drag and drop

sashaandjic commented:
First time I saw the page I though that there was something wrong. Can we improve the ux a bit?
![image](https://user-images.githubusercontent.com/23243244/115903106-34231800-a431-11eb-8b5a-75ce15dd3ca7.png)

The UI modifications to Variants uploader should be applied to the Somatic Variants form as well.
![image](https://user-images.githubusercontent.com/23243244/115902996-0e960e80-a431-11eb-983d-ad85c025f58e.png)

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Grab the subject types labels programmatically

OrdiNeu committed:
[wip] LFS-842: Improve UX in the Variants Upload form

Attempt to load everythign into a card.

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Use an alert as suggested by @marta- to display file naming convention

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+            onSubmit={upload}
+            key="file-upload"
+            id="variantForm">
+        <Grid container direction="row-reverse" justify="flex-end">
```
Looks very misaligned vertically on wide screens:
![image](https://user-images.githubusercontent.com/651980/115964119-dd7d1300-a4f0-11eb-9064-ad9882ed40e0.png)

marta- replied:
There's also a huge gap between the drop zone and the info box on wide screens. The dropzone should stretch across the whole width of its containing grid item.

marta- replied:
Addressed in 9dee4a1

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+            <Alert severity="info">
+              <AlertTitle>Expected file name format</AlertTitle>
+              <div>PatientId_TumorNumber.csv (e.g. AB12345_1.csv)</div>
+              <div>PatientId_TumorNumber_TumorRegion.csv (e.g. AB12345_1_a.csv)</div>
```
Per @sashaandjic 's suggestion:
```suggestion
              <div>Patient_Tumor.csv (e.g. AB12345_1.csv)</div>
              <div>Patient_Tumor_TumorRegion.csv (e.g. AB12345_1_a.csv)</div>
```

marta- replied:
Done in 466ffc8.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+                    { file?.region?.path && <> / {file.region?.type?.label || "Tumor Region"}: <Link href={regionPath} target="_blank"> {file.region.id} </Link> </> }
+                    { file.formPath && <span>Somatic Variants / <Link href={file.formPath.replace("/Forms", "Forms")} target="_blank"> {file.name} </Link></span> }
+                  </Typography>
+                : <span>
```
`span`s (and inline elements in general) should not contain any block elements, and `<TextFiled>` generates a `div`, which is block.

marta- replied:
Addressed in 8acd21c.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+                    className={classes.fileDetail}
+                  />
+                  <label htmlFor="contained-button-file">
+                    <Button type="submit" variant="contained" color="primary" disabled={uploadInProgress || !!error && selectedFiles.length == 0} className={classes.uploadButton} form="variantForm">
```
The button is misaligned vertically:
![image](https://user-images.githubusercontent.com/651980/116012424-616df280-a5f8-11eb-82ad-984075ecff57.png)

Also, on narrower screens, it doesn't wrap nicely:
![image](https://user-images.githubusercontent.com/651980/116012415-4bf8c880-a5f8-11eb-91ba-9b984c4eaf58.png)

marta- replied:
Addressed in 8acd21c.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
-            </div>
-        ) } ) }
-    </span> }
+                        <Typography variant="caption">
```
`<Typography>` inside `<Typography>`

marta- replied:
Addressed in 8acd21c.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
-        ) } ) }
-    </span> }
+                        <Typography variant="caption">
+                          <Link
```
`Link` can have `variant` just like `Typography`, no need to wrap one inside the other.

marta- replied:
Addressed in 8acd21c.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+                width: '1%',
+                whiteSpace: 'nowrap',
+              },
+              render: rowData => <Link to={"/content.html" + rowData["@path"]}>
```
These links don't work. Should be `href` instead of `to` and the path should not be prefixed with "/content.html".

marta- replied:
Fixed in 7344b23.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+            key="file-upload"
+            id="variantForm">
+        <Grid container direction="row-reverse" justify="flex-end">
+          <Grid item md={12} lg={6}>
```
On narrow screens, neither the drop zone nor the alert stretch to full width, making the page look messy.

![image](https://user-images.githubusercontent.com/651980/116012365-0b994a80-a5f8-11eb-96d2-824a03396341.png)

marta- replied:
Addressed in 9dee4a1.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
-              }
-              <Typography variant="body1" component="div" className={classes.fileInfo}>
-                  {(!file.sameFiles || file.sameFiles.length == 0)
+                <Typography variant="body1" component="div" className={classes.fileInfo}>
```
There's an unnecessary padding around the version info text, and the left indentation makes it look misaligned:

![image](https://user-images.githubusercontent.com/651980/116012482-b27de680-a5f8-11eb-9afa-10b3aa072e88.png)

![image](https://user-images.githubusercontent.com/651980/116012507-cde8f180-a5f8-11eb-8dfd-dbe9ce96f636.png)

marta- replied:
Addressed in 8acd21c.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +696,181 @@ export default function VariantFilesContainer() {
+              render: rowData => <Tooltip title={"Download"}>
+                                  <IconButton>
+                                    <GetApp />
+                                  </IconButton>
```
There's no functionality attached to this button.

marta- replied:
Fixed in 7344b23.

marta- commented:
> The UI modifications to Variants uploader should be applied to the Somatic Variants form as well.

Which UI modifications did you mean?
* The file name format isn't required if uploading directly to a Somatic Variants form. We only need it on the Variants Upload page to determine what subject to assign to. So, the message about how the file should be named, and any UI displaying the editable subject id are not useful on the form.
* The upload button is also unnecessary, on the form we have "Save" which does the upload.
* The other versions of the uploaded file that we mention on Variant Upload are really files attached to other Somatic Variant forms but named in the same way, so again probably not relevant while in the form.

marta- committed:
LFS-842: Improve UX in the Variants Upload form
Fix the layout of the drop zone

marta- committed:
LFS-842: Improve UX in the Variants Upload form
* Fix upload button alignment
* Fix version info alignment
* Fix invalid html markup
* Simplify React markup

marta- committed:
LFS-842: Improve UX in the Variants Upload form
Fixed links and actions of "other versions" of the uploaded file

marta- committed:
LFS-842: Improve UX in the Variants Upload form
Improve message about expected file formats

marta- committed:
LFS-842: Improve UX in the Variants Upload form
[cleanup] Removed debug code

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Switch the title of the questionnaire from being hard-coded to
retrieving it on page load.

Fix the spacing and positioning of the seperators for newly updated form
subjects.

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -169,6 +169,7 @@ export default function VariantFilesContainer() {
   // uuids of the Subjects and the SomaticVariants questionnaire
   // To be fetch on page load
   let [ somaticVariantsUUID, setSomaticVariantsUUID ] = useState();
+  let [ somaticVariantsTItle, setSomaticVariantsTitle ] = useState("");
```
Typo: `TItle`

OrdiNeu committed:
LFS-842: Improve UX in the Variants Upload form

Typofix

marta- committed:
LFS-842: Improve UX in the Variants Upload form
* Another markup fix
* Better path display for the uploaded file

marta- committed:
LFS-1056 Inconsistency when uploading multiple files in the Variants Upload form
Temporarily ignore all but the first selected file until concurrency issues are solved

sashaandjic reviewed (approved):
tested

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -660,118 +715,179 @@ export default function VariantFilesContainer() {
+                  <Typography variant="overline" component="div" className={classes.fileDetail}>
+                    {file.subject?.type?.label || "Patient"} <Link href={subjectPath} target="_blank"> {file.subject.id} </Link> /&nbsp;
+                    {file.tumor?.type?.label || "Tumor"} <Link href={tumorPath} target="_blank"> {file.tumor.id} </Link>
+                    { file?.region?.path && <> / {file.region?.type?.label || "Tumor Region"}: <Link href={regionPath} target="_blank"> {file.region.id} </Link> </> }
```
```suggestion
                    { file?.region?.path && <> / {file.region?.type?.label || "Tumor Region"} <Link href={regionPath} target="_blank"> {file.region.id} </Link> </> }
```

marta- committed:
LFS-842: Improve UX in the Variants Upload form

Removed inconsistent separator

sdumitriu reviewed (commented):
Not sure if this is a regression or an existing bug:

- select `p_t.csv` for upload, then enter a value in the region input; upload -> region is ignored, file is uploaded to `p / t`
- select `p_t_r.csv` for upload, then delete the value for r, upload -> a region with no name is created, with weird display (nothing/empty in tables and on the form, `undefined` in the patient chart)

sdumitriu reviewed (approved):
Old bugs, to be addressed later.
